File: CCIP/CCIP/Class+addition/UIView+addition.m
//
//  UIView+DashedLine.m
//  CCIP
//
//  Created by 腹黒い茶 on 2017/07/23.
//  Copyright © 2017年 CPRTeam. All rights reserved.
//

#import "UIView+addition.h"

@implementation UIView (DashedLine)

- (void)addDashedLine:(UIColor *)color {
    for (CALayer *layer in self.layer.sublayers) {
        if ([layer.name isEqualToString:DASHLINE_VIEW_ID]) {
            [layer removeFromSuperlayer];
        }
    }
    [self setBackgroundColor:[UIColor clearColor]];
    
    CAShapeLayer *shapeLayer = [CAShapeLayer layer];
    [shapeLayer setName:DASHLINE_VIEW_ID];
    [shapeLayer setBounds:self.bounds];
    [shapeLayer setPosition:CGPointMake(self.frame.size.width / 2, self.frame.size.height / 2)];
    [shapeLayer setFillColor:[[UIColor clearColor] CGColor]];
    [shapeLayer setStrokeColor:color.CGColor];
    [shapeLayer setLineWidth:self.frame.size.height];
    [shapeLayer setLineJoin:kCALineJoinRound];
    [shapeLayer setLineDashPattern:@[ @(5), @(5) ]];
    
    CGAffineTransform transform = self.transform;
    CGMutablePathRef path = CGPathCreateMutable();
    CGPathMoveToPoint(path, &transform, 0, 0);
    CGPathAddLineToPoint(path, &transform, self.frame.size.width, 0);
    [shapeLayer setPath:path];
    
    [self.layer addSublayer:shapeLayer];
}

@end

@implementation UIView (linear_diagonal_gradient)

- (void)sizeGradientToFit {
    [self setGradientColor:nil
                        To:nil
                StartPoint:CGPointZero
                   ToPoint:CGPointZero];
}

- (void)setGradientColor:(UIColor * _Nullable)from To:(UIColor * _Nullable)to StartPoint:(CGPoint)fromPoint ToPoint:(CGPoint)toPoint {
    NSString *name = @"GradientBackground";
    // Set view background linear diagonal gradient
    //   Create the gradient
    CAGradientLayer *theViewGradient = nil;
    for (CALayer *layer in [self.layer sublayers]) {
        if (layer.name == name) {
            theViewGradient = (CAGradientLayer *)layer;
            [layer removeFromSuperlayer];
            break;
        }
    }
    if (theViewGradient == nil) {
        theViewGradient = [CAGradientLayer layer];
        theViewGradient.name = name;
        theViewGradient.frame = CGRectMake(0, 0, self.frame.size.width, self.frame.size.height);
    }
    if (from != nil && to != nil && !CGPointEqualToPoint(CGPointZero, fromPoint) && !CGPointEqualToPoint(CGPointZero, toPoint)) {
        theViewGradient.colors = [NSArray arrayWithObjects: (id)from.CGColor, (id)to.CGColor, nil];
        theViewGradient.startPoint = fromPoint;
        theViewGradient.endPoint = toPoint;
    } else {
        theViewGradient.frame = CGRectMake(0, 0, self.frame.size.width, self.frame.size.height);
    }
    [self.layer insertSublayer:theViewGradient
                       atIndex:0];
}

@end

File: CCIP/CCIP/Class+addition/UIView+addition.swift
//
//  UIView+addition.swift
//  OPass
//
//  Created by 腹黒い茶 on 2018/11/4.
//  Copyright © 2018 OPass. All rights reserved.
//

import Foundation
import UIKit

@objc extension UIView { // DashedLine and linear diagonal gradient
    func addDashedLine(_ color : UIColor) {
        for layer in self.layer.sublayers! {
            if (layer.name == Constants.DashlineViewId) {
                layer.removeFromSuperlayer();
            }
        }
        self.backgroundColor = UIColor.clear;

        let shapeLayer : CAShapeLayer = CAShapeLayer.init();
        shapeLayer.name = Constants.DashlineViewId;
        shapeLayer.bounds = self.bounds;
        shapeLayer.position = CGPoint.init(x: self.frame.size.width / 2, y: self.frame.size.height / 2);
        shapeLayer.fillColor = UIColor.clear.cgColor;
        shapeLayer.strokeColor = color.cgColor;
        shapeLayer.lineWidth = self.frame.size.height;
        shapeLayer.lineJoin = .round;
        shapeLayer.lineDashPattern = [ 5, 5 ];

        let transform : CGAffineTransform = self.transform;
        let path : CGMutablePathRef = CGPathCreateMutable();
        CGPathMoveToPoint(path, &transform, 0, 0);
        CGPathAddLineToPoint(path, &transform, self.frame.size.width, 0);
        shapeLayer.path = path;

        self.layer.addSublayer(shapeLayer);
    }

    func setGradientColor(_ from : UIColor, to: UIColor, startPoint:CGPoint, toPoint:CGPoint) {
        var name : String = "GradientBackground";
        // Set view background linear diagonal gradient
        //   Create the gradient
        let theViewGradient : CAGradientLayer? = nil;
        for layer in self.layer.sublayers! {
            if (layer.name == name) {
                theViewGradient = layer as CAGradientLayer;
                layer.removeFromSuperlayer();
                break;
            }
        }
        if (theViewGradient == nil) {
            theViewGradient = CAGradientLayer.layer;
            theViewGradient.name = name;
            theViewGradient.frame = CGRect(x: 0, y: 0, width: self.frame.size.width, height: self.frame.size.height);
        }
        if (from != nil && to != nil && CGPointZero != startPoint && CGPointZero != toPoint) {
            theViewGradient.colors = [ from.CGColor, to.CGColor ];
            theViewGradient.startPoint = startPoint;
            theViewGradient.endPoint = toPoint;
        } else {
            theViewGradient.frame = CGRect(x: 0, y: 0, width: self.frame.size.width, height: self.frame.size.height);
        }
        self.layer.insertSublayer(theViewGradient, at: 0);
    }

    func sizeGradientToFit() {
        self.setGradientColor(nil, to: nil, startPoint: CGPoint.zero, toPoint: CGPoint.zero);
    }
}
